pennylane/compiler/python_compiler/transforms/quantum/tree_traversal.py

```diff
@@ -0,0 +1,1587 @@
+                case quantum.CustomOp():
+                    for i, qb in enumerate(chain(op.in_qubits, op.in_ctrl_qubits)):
+                        qubit_to_reg_idx[op.out_qubits[i]] = i
+                        qubit_to_reg_idx[op.results[i]] = qubit_to_reg_idx[qb]
+                        del qubit_to_reg_idx[qb]
+                case quantum.InsertOp():
+                    assert qubit_to_reg_idx[op.qubit] is op.idx_attr if op.idx_attr else True
+                    del qubit_to_reg_idx[op.qubit]
+                    # update register since it might have changed
+                    op.operands = (current_reg, op.idx, op.qubit)
+                    current_reg = op.out_qreg
+                case quantum.MeasureOp():
+                    # find the qubit to be measured and its index
+                    mcm_qubit, mcm_idx = next(
+                        ((qb, idx) for qb, idx in qubit_to_reg_idx.items() if qb == op.in_qubit),
+                        (None, None),
+                    )
+                    if mcm_qubit is None:
+                        raise RuntimeError(
+                            f"Could not find qubit {op.in_qubit} in register mapping"
+                        )
+
+                    insert_ops = set()
+
+                    # create a register boundary before the measure
+                    rewriter.insertion_point = InsertPoint.before(op)
+                    for qb, idx in qubit_to_reg_idx.items():
+                        insert_op = quantum.InsertOp(current_reg, idx, qb)
+                        rewriter.insert(insert_op)
+                        insert_ops.add(insert_op)
+                        current_reg = insert_op.out_qreg
+
+                    # extract the qubit that will be measured from the register
+                    extract_op = quantum.ExtractOp(current_reg, mcm_idx)
+                    extract_op.attributes["meas_boundary"] = builtin.UnitAttr()
+                    rewriter.insert(extract_op)
+                    # update the measure operation to use the extracted qubit
+                    op.operands = (extract_op.qubit,)
+
+                    # update the map to process the measure qubit - remove the old qubit
+                    del qubit_to_reg_idx[mcm_qubit]
+
+                    # restore qubit values from before the register boundary
+                    rewriter.insertion_point = InsertPoint.after(op)
+
+                    def _create_replace_filter(excluded_operations):
+                        """Create a filter function to avoid cell-var-from-loop issues."""
+                        return lambda use: use.operation not in excluded_operations
+
+                    for qb, idx in list(qubit_to_reg_idx.items()):
+                        extract_op = quantum.ExtractOp(current_reg, idx)
+                        rewriter.insert(extract_op)
+                        qb.replace_by_if(extract_op.qubit, _create_replace_filter(insert_ops))
```

Comment:
Why not just do this? The `_create_replace_filter` function just adds complexity.
```suggestion
                        qb.replace_by_if(extract_op.qubit, lambda use: use.operation not in insert_ops)
```